src/Compiler/Service/FSharpProjectSnapshot.fs

```diff
@@ -675,27 +675,32 @@ and [<Experimental("This FCS API is experimental and subject to change.")>] FSha
             |> async.Return
 
         FSharpProjectSnapshot.FromOptions(options, getFileSnapshot)
+        
+let internal snapshotTable = ConditionalWeakTable<ProjectSnapshot, FSharpProjectOptions>()
+
+let rec internal snapshotToOptions (projectSnapshot: ProjectSnapshot) =
+    snapshotTable.GetValue(projectSnapshot, fun projectSnapshot ->
+        {
+            ProjectFileName = projectSnapshot.ProjectFileName
+            ProjectId = projectSnapshot.ProjectId
+            SourceFiles = projectSnapshot.SourceFiles |> Seq.map (fun x -> x.FileName) |> Seq.toArray
+            OtherOptions = projectSnapshot.CommandLineOptions |> List.toArray
+            ReferencedProjects =
+                projectSnapshot.ReferencedProjects
+                |> Seq.map (function
+                    | FSharpReference(name, opts) -> FSharpReferencedProject.FSharpReference(name, opts.ProjectSnapshot |> snapshotToOptions)
+                    | PEReference(getStamp, reader) -> FSharpReferencedProject.PEReference(getStamp, reader)
+                    | ILModuleReference(name, getStamp, getReader) -> FSharpReferencedProject.ILModuleReference(name, getStamp, getReader))
+                |> Seq.toArray
+            IsIncompleteTypeCheckEnvironment = projectSnapshot.IsIncompleteTypeCheckEnvironment
+            UseScriptResolutionRules = projectSnapshot.UseScriptResolutionRules
+            LoadTime = projectSnapshot.LoadTime
+            UnresolvedReferences = projectSnapshot.UnresolvedReferences
+            OriginalLoadReferences = projectSnapshot.OriginalLoadReferences
+            Stamp = projectSnapshot.Stamp
+        }    
```

Comment:
This is the main fix around creating too many options

Reply:
This makes sense. Out of curiosity what are you using the resulting options for?

Reply:
There seems to be paths that still want a `FSharpProjectOptions`.  Following the current use of `snapshotToOptions` should show them.

Specifically they show up in transparent compiler which get turned into longer lived objects here:

https://github.com/dotnet/fsharp/blob/f4860a40909b9da3e42f160d7a6205872ced737c/src/Compiler/Service/TransparentCompiler.fs#L1629 
https://github.com/dotnet/fsharp/blob/f4860a40909b9da3e42f160d7a6205872ced737c/src/Compiler/Service/TransparentCompiler.fs#L1898

Reply:
Ah, I see, didn't notice these usages. Wonder what would be a good way to phase them out of there... But for now at least your fix improves it.

Reply:
Yeah that would the real fix. I did do before that but it was more invasive.